frontend/src/scenes/trends/ActionFilter/ActionFilter.js

```diff
@@ -3,9 +3,9 @@ import { useActions, useValues } from 'kea'
 import { entityFilterLogic } from './actionFilterLogic'
 import { ActionFilterRow } from './ActionFilterRow'
 
-export function ActionFilter(props) {
-    const { allFilters, filters } = useValues(entityFilterLogic)
-    const { createNewFilter, initializeLocalFilters } = useActions(entityFilterLogic)
+export function ActionFilter({ setFilters, defaultFilters, showMaths }) {
+    const { allFilters, filters } = useValues(entityFilterLogic({ setFilters, defaultFilters }))
+    const { createNewFilter, initializeLocalFilters } = useActions(entityFilterLogic({ setFilters, defaultFilters }))
```

Comment:
I haven't written about it in the docs, but seeing this I get a feeling that passing `setFilters`, which in the end is just a kea action (or sometimes `(filters) => setFunnel({ filters })`?), as `props` to the logic seems like there's a lot of entanglement happening here and it's hard to follow along when reading the code.

It could be perfectly fine though. The mental model I usually look for is this:
- if everything follows a *straight line* from react via an action to the logic to its reducers/selectors and then back to react via a value... it's good
- if there's *loop* in the line, either by using react as a conveyor belt to move actions around different logics or using react as something that reacts to value changes and then dispatches new actions (like the useEffect a few lines below)... it smells fishy and can probably be simplified

Reply:
Managed to do everything else but struggled with this one.

I guess at the moment it's more 2. The filters go from a funnel -> ActionFilter (where a user can add/remove actions/events) -> back to funnel to be stored/saved.

ActionFilter takes a filter dict as input, and returns a changed filter dict as output to the setFilter function. This does run via react, but because it's in 2 places (funnels and trends) I struggled to work out how to only do this via kea. Do you have any suggestions there?

Reply:
I thought I'd chime in here as the actionfilter is giving some difficulty. 
1. The reason the action filters duplicates the original filters it's passed in is to preserve the ordering when users create the filters because we have multiple now. If we re-rendered them based on the dict it wouldn't always place filters as the last one. 
2. The useeffect was a big edge case that I didn't want to use but placed in there to solve the instance upon initial load where the parent trendlogic hadn't retrieved any filters yet but gave the actionfilter logic an empty filter array. 

Would definitely prefer not to use it, but wasn't fully sure how to set up the entitylogic to wait for the initial load of the filters to complete

Reply:
I think I've managed to make this work. Basically all `filters` is at the moment is just a selectors that takes `allFilters` and puts it into a dict that looks like `{actions: [], events: []}`. This massively simplified the Logic everywhere, and also allowed me (I think!) to get rid of useEffect.

Reply:
@EDsCODE yeah, there's no good "wait until x is available" pattern that I know of, other than just using a listener to update the values once they do become available.

It *might* be worth looking into adding a new event, `onPropsChanged` or something like that which will get called every time the props are updated. However that might be more of a hack than a solution, I don't know :).

Tim's last commits did simplify a lot and hopefully removed the need for something like this entirely.